### PR TITLE
Bug 1139314 - [Stingray] let focus manager to setTimeout and call focus ...

### PR DESCRIPTION
**Pull request by huchengtw-moz**
...of top most UI

### CODE REVIEW

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=fcdb5b9af94b)

File: tv_apps/smart-system/js/focus_manager.js [hunk 1]
```diff
@@ -29,29 +29,41 @@
...
-    if (visible.length > 0) {
-      // sort system UIs
-      visible.sort(function(item1, item2) {
-        return item2.getOrder() - item1.getOrder();
+    // According to gecko's implementation, we cannot change focus while
+    // handling keyboard event. The way to bypass it is use setTimeout to lazy
+    // run the focus switching.
+    //
+    // There is another focus rule at gecko: we cannot move focus cross domain,
+    // like to move focus from a button is remote iframe to system app's button.
```
Comment (luke-chang) [inline tv_apps/smart-system/js/focus_manager.js]: There is another focusing rule in gecko: we cannot move focus across domains, like moving focus from a button in a remote iframe to another one in system app.

File: tv_apps/smart-system/js/focus_manager.js [hunk 2]
```diff
@@ -29,29 +29,41 @@
...
-    // list all visible system UI
-    var visible = systemUIList.filter(function(item) {
-      return item && item.isVisible();
-    });
-    if (visible.length > 0) {
-      // sort system UIs
-      visible.sort(function(item1, item2) {
-        return item2.getOrder() - item1.getOrder();
+    // According to gecko's implementation, we cannot change focus while
+    // handling keyboard event. The way to bypass it is use setTimeout to lazy
```
Comment (luke-chang) [inline tv_apps/smart-system/js/focus_manager.js]: is "using"

File: tv_apps/smart-system/js/focus_manager.js [hunk 3]
```diff
@@ -29,29 +29,41 @@
...
+    // run the focus switching.
+    //
+    // There is another focus rule at gecko: we cannot move focus cross domain,
+    // like to move focus from a button is remote iframe to system app's button.
+    // The way to bypass it is use 'document.activeElement.blur()'. But we don't
```
Comment (luke-chang) [inline tv_apps/smart-system/js/focus_manager.js]: is "using"

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a90b9027f6e1)

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d6e3c82d7247)

File: tv_apps/smart-system/js/focus_manager.js [hunk 4]
```diff
@@ -29,29 +29,41 @@
...
-      // focus top-most system UI
-      visible[0].focus();
-    } else {
-      // no system UI, we set focus back to top-most AppWindow
-      AppWindowManager.getActiveApp().focus();
-    }
-
+      if (visible.length > 0) {
+        // sort system UIs
+        visible.sort(function(item1, item2) {
```
Comment (luke-chang) [inline tv_apps/smart-system/js/focus_manager.js]: Do we need a stable sort here?

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=bc86f9cfa406)

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0411ad4b00f1)

**Comment by try-server-hook** [comment]
[huchengtw-moz John Hu (huchengtw-moz)](https://github.com/huchengtw-moz) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=056fbae344c9)

**Pull request merged**